pkg/mcs/meta_storage/server/grpc_service.go

```diff
@@ -0,0 +1,217 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/pingcap/kvproto/pkg/meta_storagepb"
+	bs "github.com/tikv/pd/pkg/basicserver"
+	"github.com/tikv/pd/pkg/mcs/registry"
+	"github.com/tikv/pd/pkg/utils/apiutil"
+	"go.etcd.io/etcd/clientv3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	// errNotLeader is returned when current server is not the leader.
+	errNotLeader = status.Errorf(codes.Unavailable, "not leader")
+)
+
+var _ meta_storagepb.MetaStorageServer = (*Service)(nil)
+
+// SetUpRestHandler is a hook to sets up the REST service.
+var SetUpRestHandler = func(srv *Service) (http.Handler, apiutil.APIServiceGroup) {
+	return dummyRestService{}, apiutil.APIServiceGroup{}
+}
+
+type dummyRestService struct{}
+
+func (d dummyRestService) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusNotImplemented)
+	w.Write([]byte("not implemented"))
+}
+
+// Service is the gRPC service for meta storage.
+type Service struct {
+	ctx     context.Context
+	manager *Manager
+	// settings
+}
+
+// NewService creates a new meta storage service.
+func NewService[T ClusterIDProvider](svr bs.Server) registry.RegistrableService {
+	return &Service{
+		ctx:     svr.Context(),
+		manager: NewManager[T](svr),
+	}
+}
+
+// RegisterGRPCService registers the service to gRPC server.
+func (s *Service) RegisterGRPCService(g *grpc.Server) {
+	meta_storagepb.RegisterMetaStorageServer(g, s)
+}
+
+// RegisterRESTHandler registers the service to REST server.
+func (s *Service) RegisterRESTHandler(userDefineHandlers map[string]http.Handler) {
+	handler, group := SetUpRestHandler(s)
+	apiutil.RegisterUserDefinedHandlers(userDefineHandlers, &group, handler)
+}
+
+func (s *Service) checkServing() error {
+	if !s.manager.srv.IsServing() {
+		return errNotLeader
+	}
+	return nil
+}
+
+// Watch watches the key with a given prefix and revision.
+func (s *Service) Watch(req *meta_storagepb.WatchRequest, server meta_storagepb.MetaStorage_WatchServer) error {
+	if err := s.checkServing(); err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(s.ctx)
+	defer cancel()
+	key := string(req.GetKey())
+	endKey := string(req.GetRangeEnd())
+	startRevision := req.GetStartRevision()
+	cli := s.manager.GetClient()
+	watchChan := cli.Watch(ctx, key, clientv3.WithPrefix(), clientv3.WithRange(endKey), clientv3.WithRev(startRevision), clientv3.WithPrevKV())
```

Comment:
Do we need to let the client can pass the options? such as some clients may no need `WithPrevKV`.

Reply:
Done, and remove with prefix since it should be handled in the client side.